### DISCUSSION TITLE
How to tell mise to some python packages without using a venv or global ~/.default-python-packages?

### DISCUSSION THREAD

**Discussion by ssbarnea** [jdx/mise#3875] [Q&A]
I do want to install some python packages like `uv` and `tox` but without making use of the global option `~/.default-python-packages` because this one cannot be configured inside the repository and it does affect other tools.

I know that the first suggestion would be to make use of `venv` but for some projects this would be an overkill.

Is there another way to tell mise to use a python version end ensure that a list of pip dependencies are also installed? -- Please note that I do not want to run another command to do this manually.

To better explain myself, I use `tox` to test over 95% python projects and tox takes care of virtualenv management itself. Still, installing tox on a new machine (CI too) would require additional steps, so it would be much better if I can just test mise to take care of this and only mention mise as a required build tool, reducing the number of steps require to prepare the development/testing environment.

Apparently the https://mise.jdx.dev/lang/python.html#default-python-packages option is one of the few that does not have a mise.toml configuration option and can only be controlled via `MISE_PYTHON_DEFAULT_PACKAGES_FILE`,.. not really sure why.

Update: I also observed that if I modify the `.default-python-packages` file mise to add something to it, mise will not know to install the missing tools and using the `mise install python --force` would force full download of the entire python. Apparently no middle ground for updating these dependencies.

**Comment by syhol** [comment]
You're not the first to want this behavior. Others have solved it by using a postinstall hook to call something like `uv pip install -r tools.txt`. You don't need to call any other commands and it will run on a mise install.

You could also do it on an enter hook, but that requires mise to be activated. Also it will run each time you enter the directory, not just the first time, so that could get annoying after a while.

**Comment by jubr** [comment]
I got mad & figured it out.
This works by entering the dir and does not depend on `mise install` being run, which `postinstall` solutions do require.

```toml
[tools]
uv     = "0.11.26"
python = "3.13.9"    # Mise manages python, also for uv

[env]
VENV = ".venv" # DRY
# Automatically create and activate Python `.venv`
_.python.venv = { path = "{{ env.VENV }}", create = true, uv_create_args = ['--seed'] }
# Prefer `.venv` python, so packages are installed there. Else ensure mise's version.
UV_PYTHON = { tools = true, value = '''
  {%- set py = [config_root, env.VENV, 'bin', 'python'] | join_path -%}
  {%- if py is exists -%}
    {{ py }}
  {%- elif tools.python is defined and tools.python.path is defined -%}
    {{ tools.python.path }}
  {%- else -%}
    python
  {%- endif -%}
''' }
UV_PYTHON_DOWNLOADS = "never"
# With `tools = true` late evaluation, this is after `.venv` creation
VIRTUAL_ENV_AUTOINSTALL_PACKAGES = { tools = true, value = '''
  {%- set vv = '' -%}
  {%- if tools.python is defined and tools.python.version is defined -%}
    {%- set vv = tools.python.version | split(pat='.') | slice(end=2) | join(sep='.') -%}
  {%- endif -%}
  {%- if vv == '' -%}
    unknown-python-version
  {%- else -%}
    {%- set_global first_req = '' -%}
    {%- for line in read_file(path='requirements.txt') | split(pat='\n') -%}
      {%- set trimmed = line | trim -%}
      {%- if trimmed != '' and trimmed is not starting_with('#') -%}
        {%- set_global first_req = trimmed -%}
        {%- break -%}
      {%- endif -%}
    {%- endfor -%}
    {%- set pkg = first_req | replace(from='=', to='|') | replace(from='!', to='|') | replace(from='~', to='|')
                            | replace(from='<', to='|') | replace(from='>', to='|') | split(pat='|') | first | trim -%}
    {%- set tst = [config_root, env.VENV, 'lib', 'python' ~ vv, 'site-packages', pkg] | join_path -%}
    {%- if tst is exists -%}
      installed
    {%- else -%}
      {{- exec(command='set -x; uv pip install -r requirements.txt') -}}
      installed-just-now
    {%- endif -%}
  {%- endif -%}
''' }

# Don't need this one
#[hooks]
#postinstall = "uv pip install -r requirements.txt"
```

Which will look something like:
```bash
$ rm -rf .venv/

mise creating venv with uv at: ~/path/to/project/.venv
Using CPython 3.13.9 interpreter at: /opt/mise/installs/python/3.13.9/bin/python
Creating virtual environment with seed packages at: .venv
 + pip==26.1.2
Activate with: source .venv/bin/activate
+ uv pip install -r requirements.txt
Resolved 1 package in 6ms
Installed 1 package in 55ms
 + numpy==2.5.0
```

#Enjoy.

_Update_: added guards around accessing `tools.python.version` when it is not yet installed

**Reply by syhol** [reply]
Also take a look at https://mise.jdx.dev/dev-tools/deps.html. it doesn't auto run on cd, but you might be able to set something up.
